### PR TITLE
feat(api): introduce route.fetch and route.fulfill(json)

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: docs/src/api/class-route.md [hunk 1]
```diff
@@ -386,28 +386,115 @@ affect the route matching, all the routes are matched using the original request
...
+```
+
+```java
+page.route("https://dog.ceo/api/breeds/list/all", route -> {
+  APIResponse response = route.fetch();
+  JsonObject json = new Gson().fromJson(response.text(), JsonObject.class);
+  json.set("big_red_dog", new JsonArray());
+  route.fulfill(new Route.FulfillOptions()
+    .setResponse(response)
+    .setJson(json.toString()));
```
Comment (yury-s) [inline docs/src/api/class-route.md:436]: `setBody` ? there is no point in setJson in java since we don't support any particular json lib.

File: docs/src/api/class-route.md [hunk 2]
```diff
@@ -386,28 +386,115 @@ affect the route matching, all the routes are matched using the original request
...
+page.route("https://dog.ceo/api/breeds/list/all", handle)
+```
+
+```csharp
+await page.RouteAsync("https://dog.ceo/api/breeds/list/all", async route =>
+{
+    var response = await route.FetchAsync();
+    dynamic json = await response.JsonAsync();
+    json.big_red_dog = new string[] {};
+    await route.FulfillAsync(response: response, json: json);
```
Comment (dgozman) [inline docs/src/api/class-route.md:466]:
```suggestion
    await route.FulfillAsync(new() { Response = response, Json = json });
```

File: docs/src/api/class-route.md (hunk 1 above)
Comment (dgozman) [inline docs/src/api/class-route.md:436]:
```suggestion
    .setBody(json.toString()));
```

File: docs/src/api/class-route.md [hunk 3]
```diff
@@ -519,6 +606,57 @@ Optional response body as text.
...
+
+JSON response. This method will set the content type to `application/json` if not set.
+
+**Usage**
+
+```csharp
+await page.RouteAsync("https://dog.ceo/api/breeds/list/all", async route =>
+{
+    var json = /* JsonElement with test payload */;
+    await route.FulfillAsync(json: json);
```
Comment (dgozman) [inline docs/src/api/class-route.md:656]:
```suggestion
    await route.FulfillAsync(new() { Json = json });
```

File: packages/playwright-core/src/client/network.ts [hunk 4]
```diff
@@ -307,18 +307,31 @@ export class Route extends ChannelOwner<channels.RouteChannel> implements api.Ro
...
       this._reportHandled(true);
     });
   }
 
-  private async _innerFulfill(options: { response?: api.APIResponse, status?: number, headers?: Headers, contentType?: string, body?: string | Buffer, path?: string } = {}): Promise<void> {
+  private async _innerFulfill(options: { response?: api.APIResponse, status?: number, headers?: Headers, contentType?: string, body?: string | Buffer, json?: any, path?: string } = {}): Promise<void> {
     let fetchResponseUid;
     let { status: statusOption, headers: headersOption, body } = options;
 
+    if (options.json) {
```
Comment (dgozman) [inline packages/playwright-core/src/client/network.ts:330]: To allow empty strings, zeroes, and other falsy values:

```suggestion
    if (options.json !== undefined) {
```

File: packages/playwright-core/src/client/network.ts [hunk 5]
```diff
@@ -307,18 +307,31 @@ export class Route extends ChannelOwner<channels.RouteChannel> implements api.Ro
...
+      assert(options.body === undefined, 'Can specify either body or json parameters');
```
Comment (yury-s) [inline packages/playwright-core/src/client/network.ts:331]: nit: can specify only one of

File: packages/playwright-core/src/client/network.ts [hunk 6]
```diff
@@ -351,6 +364,8 @@ export class Route extends ChannelOwner<channels.RouteChannel> implements api.Ro
       headers[header.toLowerCase()] = String(headersOption![header]);
     if (options.contentType)
       headers['content-type'] = String(options.contentType);
+    if (options.json)
```
Comment (dgozman) [inline packages/playwright-core/src/client/network.ts:367]: If I explicitly pass `contentType` option, we should not override with `application/json`.

File: tests/library/browsercontext-fetch.spec.ts [hunk 7]
```diff
@@ -797,11 +797,6 @@ it('should dispose when context closes', async function({ context, server }) {
   expect(error.message).toContain('Response has been disposed');
 });
 
-it('should throw on invalid first argument', async function({ context }) {
```
Comment (yury-s) [inline tests/library/browsercontext-fetch.spec.ts:800]: Why was it removed?

File: packages/playwright-core/src/client/network.ts [hunk 8]
```diff
@@ -307,18 +307,31 @@ export class Route extends ChannelOwner<channels.RouteChannel> implements api.Ro
     this._reportHandled(true);
   }
 
-  async fulfill(options: { response?: api.APIResponse, status?: number, headers?: Headers, contentType?: string, body?: string | Buffer, path?: string } = {}) {
+  async fetch(options: FallbackOverrides = {}) {
+    this._checkNotHandled();
```
Comment (dgozman) [inline packages/playwright-core/src/client/network.ts:311]: Do we need this check? We might as well fetch after the route is handled?

File: tests/library/browsercontext-fetch.spec.ts [hunk 9]
```diff
@@ -797,11 +797,6 @@ it('should dispose when context closes', async function({ context, server }) {
...
-  const error = await context.request.get({} as any).catch(e => e);
-  expect(error.message).toContain('First argument must be either URL string or Request');
```
Comment (dgozman) [inline tests/library/browsercontext-fetch.spec.ts:802]: I see this error in the code, why remove the test?

File: tests/page/page-request-intercept.spec.ts [hunk 10]
```diff
@@ -194,4 +192,43 @@ it('should intercept multipart/form-data request body', async ({ page, server, a
...
+
+it('should intercept with post data override', async ({ page, server, isElectron, isAndroid }) => {
+  it.fixme(isElectron, 'error: Browser context management is not supported.');
+  it.skip(isAndroid, 'The internal Android localhost (10.0.0.2) != the localhost on the host');
+  const requestPromise = server.waitForRequest('/empty.html');
+  await page.route('**/*.html', async route => {
+    const response = await route.fetch({
+      data: { 'foo': 'bar' },
+    });
+    await route.fulfill({ response });
```
Comment (yury-s) [inline tests/page/page-request-intercept.spec.ts:229]: add a test where it it's fulfilled with overrides too

**Review by yury-s — APPROVED** [review]

File: docs/src/mock-js.md [hunk 11]
```diff
@@ -1,154 +0,0 @@
----
-id: mock
-title: "Mock APIs"
```
Comment (dgozman) [inline docs/src/mock-js.md:3]: Removal of this doc seems unrelated 😄

File: tests/library/browsercontext-fetch.spec.ts (hunk 7 above)
Comment (pavelfeldman) [reply tests/library/browsercontext-fetch.spec.ts:800]: We don't generally check parameter types at runtime - we have 1000+ parameters and only check the type here.

File: tests/page/page-request-intercept.spec.ts (hunk 10 above)
Comment (pavelfeldman) [reply tests/page/page-request-intercept.spec.ts:229]: Yeah, I'm not sure why it was not covered, do you want to look at it?

File: packages/playwright-core/src/client/network.ts (hunk 6 above)
Comment (pavelfeldman) [reply packages/playwright-core/src/client/network.ts:367]: I've lost 'else', the order war ok :)

File: tests/library/browsercontext-fetch.spec.ts (hunk 9 above)
Comment (pavelfeldman) [reply tests/library/browsercontext-fetch.spec.ts:802]: Because we have 1000 more places where we don't tolerate poor types?

File: packages/playwright-core/src/client/network.ts (hunk 8 above)
Comment (pavelfeldman) [reply packages/playwright-core/src/client/network.ts]: Sure, I can remove it.
